### ISSUE TITLE
Kestrel should add "Connection: close" header to custom HTTP/1.1 responses if the max request body size is exceeded

### ISSUE THREAD

**Issue by halter73** [dotnet/aspnetcore#31169]
We should consider automatically injecting a "Connection: close" header when the app writes a custom response and the max request body size has been exceeded. Without this header, the client might try to reuse the connection causing request errors.

We could do this any time Kestrel flushes headers and already knows the request body exceeds the limit. It might not be possible to know yet for some chunked requests if the body hasn't been read yet.

Do we need to worry about the app changing the limit after flushing headers? That seems unlikely but technically possible.

See also: https://github.com/dotnet/aspnetcore/issues/31136

**Comment by Tratcher** [comment]
> Without this header, the client might try to reuse the connection causing request errors.

What kind of errors?

https://tools.ietf.org/html/rfc7231#section-6.5.11
> The server MAY close the connection to prevent the client from continuing the request.

Closing the connection is optional, we shouldn't force that on app generated 413s.

**Comment by davidfowl** [comment]
I don't like injecting headers here feels magical and wrong

**Comment by halter73** [comment]
> What kind of errors?

Socket errors. We have observed this with HttpClient. It tries to reuse the connection after the 413 response and doesn't "see" the socket close/error until after it's sent the subsequent request on the same connection.

> Closing the connection is optional, we shouldn't force that on app generated 413s.

The "Connection: close" header exists to avoid these socket errors. It's optional, but it *should* be sent if the server plans to immediately close the connection.  I'm not suggesting setting the header *because* the app responded with a 413, but only because the request body size has been exceeded and we know (or can very reasonably assume) that Kestrel will close the connection.

> I don't like injecting headers here feels magical and wrong

Kestrel already injects both "Connection: close" and "Connection: keep-alive" into custom responses when it things it's appropriate. This is no more magical than that.

https://github.com/dotnet/aspnetcore/blob/dda8967c08475c523724c392c456eb337ed0cbd9/src/Servers/Kestrel/test/InMemory.FunctionalTests/ResponseTests.cs#L1311-L1331

**Comment by davidfowl** [comment]
I don't see how this behavior is expected or desirable. I think we need to see what other servers offer here before doing anything else

**Comment by Tratcher** [comment]
Design proposal:
When kestrel hits the hard size limit and is about to throw the BadRequestException, it should first set keepalive=false on that request. That will allow the normal response logic to generate the `Connection: close` header where appropriate.

@halter73 this won't have any affect on HTTP/2, right? How much of that logic is shared?

**Comment by halter73** [comment]
> @halter73 this won't have any affect on HTTP/2, right? How much of that logic is shared?

Correct. With HTTP/2, the `_keepAlive` flag is always set to false before calling middleware anyway.

https://github.com/dotnet/aspnetcore/blob/8403d676e7ce99dd430d0a6e7668937f70235559/src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs#L645-L649

**Comment by halter73** [comment]
I like the new proposal of just setting `_keepAlive` to false as soon as we throw the exception because of this simplicity. We should do it this way.

> Do we need to worry about the app changing the limit after flushing headers? That seems unlikely but technically possible.

I want to note this makes the quoted scenario _slightly_ more likely. With the previous proposal, you could observe the exception, change the limit, then flush headers and continue using the connection. With this proposal you won't. However, I find even this a super unlikely way for the app to behave.

**Comment by KristianJakubik** [comment]
Hi, this looks interesting and pretty straightforward though. I would like to try this as my first issue :).

**Comment by Tratcher** [comment]
@KristianJakubik go ahead, let us know if you have questions.

**Comment by adityamandaleeka** [comment]
@KristianJakubik Are you still interested in working on this issue?

**Comment by KristianJakubik** [comment]
Hi, sorry didn't manage to get the time for the issue as I was hoping to have. It is up for grabs, go for it 🙂.

**Comment by lbosquett** [comment]
Hi, I was seeing the code today and found that the method `SetBadRequestState` (which is used on `catch (BadHttpRequestException)` ) already sets the flag `_keepAlive` to false.

https://github.com/dotnet/aspnetcore/blob/8403d676e7ce99dd430d0a6e7668937f70235559/src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs#L1337-L1348

So now I'm in doubt, this issue is already fixed or still happens? The scenario is only about websockets?

**Comment by halter73** [comment]
`SetBadRequestState` does set `_keepAlive` to false, but that method currently only gets called when the exception is caught instead of when it's thrown. If the app catches the exception itself and writes a response before Kestrel catches the rethrown exception, the response will be written before `_keepAlive` is set to false and `Connection: close` header will not be sent. Kestrel's request body draining logic will observe this exception even if the app doesn't rethrow it, but at this point it would be too late to change the response.

My suggestion is basically just to call `SetBadRequestState` immediately when the exception is thrown. We'd probably still want to delay the call to `SetErrorResponseException` until after the app completes since that it changes the status code.

**Comment by lbosquett** [comment]
I managed to reproduce the problem with a simple file upload and reading the body on a Controller method, I started to receive the 413 without the header `Connection: close`, after reading the stack and searching references on the code I found where it is thrown.

@halter73 
 > My suggestion is basically just to call SetBadRequestState immediately when the exception is thrown. We'd probably still want to delay the call to SetErrorResponseException until after the app completes since that it changes the status code.

I tried to imeplement your suggestion in the following way, on the method `OnReadStarting` of `Http1ContentLengthMessageBody`

https://github.com/dotnet/aspnetcore/blob/8403d676e7ce99dd430d0a6e7668937f70235559/src/Servers/Kestrel/Core/src/Internal/Http/Http1ContentLengthMessageBody.cs#L236-L242
changed the if body statements:
```csharp
if (_contentLength > maxRequestBodySize)
{
    var exception = KestrelBadHttpRequestException.GetException(RequestRejectionReason.RequestBodyTooLarge, maxRequestBodySize.GetValueOrDefault().ToString(CultureInfo.InvariantCulture));
    _context.SetBadRequestState(exception);
    throw exception;
}
```

This code started to send the `connection: Close` header to browser, but also the browser started to receive the **net::ERR_INCOMPLETE_CHUNKED_ENCODING** error (2x times)

I'm not sure, but is probably related to aditional actions of setting the field `_requestRejectedException` or followed by the call of `SetErrorResponseException` and `SetErrorResponseHeaders(ex.StatusCode)` that sets the ContentLength to 0:
https://github.com/dotnet/aspnetcore/blob/8403d676e7ce99dd430d0a6e7668937f70235559/src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs#L1294

So, I tried the approach of just setting the `_keepAlive = false;` of `HttpProtocol _context` (using a new method on HttpProtocol to set the field)  at `Http1ContentLengthMessageBody.OnReadStarting` and it just worked!

![image](https://user-images.githubusercontent.com/46539053/144707947-a722c470-edb8-4dec-a28a-e2a02608f08a.png)

Can I open a PR for this or we need further investigation/design review? what do you think?

**Comment by halter73** [comment]
> This code started to send the `connection: Close` header to browser, but also the browser started is receving the **net::ERR_INCOMPLETE_CHUNKED_ENCODING** error (2x times)
>
> I'm not sure, but is probably related to aditional actions of setting the field `_requestRejectedException` or followed by the call of `SetErrorResponseException` and `SetErrorResponseHeaders(ex.StatusCode)` that sets the ContentLength to 0:
https://github.com/dotnet/aspnetcore/blob/8403d676e7ce99dd430d0a6e7668937f70235559/src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs#L1294

This is why I mentioned that we'd probably need to still delay the call `SetErrorResponseException`. I think just setting `_keepAlive` to false is also fine though if that's simpler than splitting apart `SetBadRequestState`.

> Can I open a PR for this or we need further investigation/design review? what do you think?

Go for it! Make sure to update `AddAndCheckObservedBytes` too so chunked requests are also fixed.

https://github.com/dotnet/aspnetcore/blob/bf9225b8d7fd5f8ee855dd50407a90587b812662/src/Servers/Kestrel/Core/src/Internal/Http/MessageBody.cs#L186-L195

And add a few test cases to https://github.com/dotnet/aspnetcore/blob/a450cb69b5e4549f5515cdb057a68771f56cefd7/src/Servers/Kestrel/test/InMemory.FunctionalTests/MaxRequestBodySizeTests.cs if you can. Thanks!

**Comment by halter73** [comment]
This was fixed by #38884 way back in .NET 7 preview 1, but it looks like we forgot to close the issue. Thanks again @lbosquett.

See also https://stackoverflow.com/a/73201538/719967 for how to make 413 responses more observable to the client.
